### PR TITLE
Implementation for resource templates and resource wait/fail conditions

### PR DESCRIPTION
**Pull request by jessesuen**
This implements issue #606.

It introduces a new template type, resource, which has the following fields:
```
// ResourceTemplate is a template subtype to manipulate kubernetes resources
type ResourceTemplate struct {
	// Action is the action to perform to the resource.
	// Must be one of: create, apply, delete
	Action string `json:"action"`

	// Manifest contains the kubernetes manifest
	Manifest string `json:"manifest"`

	// WaitCondition is a label selector expression which describes the conditions
	// of the k8s resource in which it is acceptable to proceed to the following step
	WaitCondition string `json:"waitCondition,omitempty"`

	// FailCondition is a label selector expression which describes the conditions
	// of the k8s resource in which the step was considered failed
	FailCondition string `json:"failCondition,omitempty"`
}
```
It has support for "wait conditions" which block the step from proceeding until the expression evaluates true, as well as "fail conditions", which fail the step immediately if evaluated true.

### CODE REVIEW

File: api/workflow/v1alpha1/types.go [hunk 1]
```diff
@@ -295,6 +313,40 @@ type Script struct {
...
+	// Action is the action to perform to the resource.
+	// Must be one of: create, apply, delete
+	Action string `json:"action"`
+
+	// Manifest contains the kubernetes manifest
+	Manifest string `json:"manifest"`
+
+	// WaitCondition is a label selector expression which describes the conditions
+	// of the k8s resource in which it is acceptable to proceed to the following step
+	WaitCondition string `json:"waitCondition,omitempty"`
```
Comment (gaganapplatix) [inline api/workflow/v1alpha1/types.go]: Can we please rename this because this is confusing that this is the success condition

Comment (jessesuen) [reply api/workflow/v1alpha1/types.go]: Ok. the agreement seems to be: `successCondition`

File: workflow/controller/operator.go [hunk 2]
```diff
@@ -635,23 +643,26 @@ func (woc *wfOperationCtx) executeTemplate(templateName string, args wfv1.Argume
 		return err
 	}
 
-	if tmpl.Container != nil {
+	switch tmpl.GetType() {
+	case wfv1.TemplateTypeContainer:
```
Comment (shrinandj) [inline workflow/controller/operator.go]: Nice!

File: workflow/controller/workflowpod.go [hunk 3]
```diff
@@ -442,23 +433,23 @@ func (woc *wfOperationCtx) addArchiveLocation(pod *apiv1.Pod, tmpl *wfv1.Templat
 	return nil
 }
 
-// addScriptVolume sets up the shared volume between init container and main container
-// containing the template script source code
-func addScriptVolume(pod *apiv1.Pod) {
-	volName := "script"
-	scriptVol := apiv1.Volume{
+// addExecutorStagingVolume sets up a shared staging volume between the init container
+// and main container for the purpose of holding the script source code for script templates
```
Comment (shrinandj) [inline workflow/controller/workflowpod.go]: This should be "holding the script source code or resource manifest".

**Comment by jessesuen** [comment]
Updated issue #620 and filed #627. 

Cascaded deletion will not happen unless ownership references are setup. But I think both cascaded deletion, labels and annotation can be achieved if the user wants this by referencing `{{workflow.name}}` in the owner references. That said we may need to have a new variable `{{workflow.uuid}}` for them to add it to the manifest:

```
  ownerReferences:
  - apiVersion: workflows.argoproj.io
    blockOwnerDeletion: true
    kind: Workflow
    name: diamond-dag-vs522
    uid: 755c6a0a-e4b6-11e7-82ee-22d4a0368ec3
```

**Pull request merged**